src/data.rs

```diff
@@ -0,0 +1,398 @@
+use anyhow::anyhow;
+use chrono::{offset::TimeZone, DateTime, Datelike, NaiveTime};
+use chrono_tz::Europe::Kiev;
+
+#[derive(Debug)]
+pub struct Subject {
+    pub id: i64,
+    pub title: String,
+    pub group: Group,
+    pub optional: bool,
+}
+
+#[derive(Debug)]
+pub struct Meeting {
+    pub id: i64,
+    pub name: String,
+    pub group: Group,
+    pub link: String,
+}
+
+#[derive(Debug)]
+pub struct Schedule {
+    pub subject_id: i64,
+    pub day: Day,
+    pub repeat: Repeat,
+    pub slot: Slot,
+}
+
+#[derive(Debug)]
+pub struct Assigned {
+    pub meeting_id: i64,
+    pub subject_id: i64,
+}
+
+#[derive(PartialEq, Debug, Clone, Copy)]
+#[repr(u8)]
+pub enum Repeat {
+    Odd = 0b01,
+    Even = 0b10,
+    Both = 0b11,
+}
+
+impl TryFrom<&str> for Repeat {
+    type Error = anyhow::Error;
+
+    fn try_from(value: &str) -> Result<Self, Self::Error> {
+        use Repeat::*;
+
+        match value {
+            "Odd" => Ok(Odd),
+            "Even" => Ok(Even),
+            "Both" => Ok(Both),
+            other => Err(anyhow!("Not a repeat: {}", other)),
+        }
+    }
+}
+
+impl<Tz: TimeZone> From<&DateTime<Tz>> for Repeat {
+    fn from(value: &DateTime<Tz>) -> Self {
+        let time = value.with_timezone(&Kiev);
+        let is_odd = ((Datelike::day(&time) + 6) / 7 % 2) == 0;
+        if is_odd {
+            Repeat::Odd
+        } else {
+            Repeat::Even
+        }
+    }
+}
+
+#[derive(PartialEq, Debug)]
+pub enum Group {
+    K25,
```

Comment:
So, are groupes hardcoded now or not?

Reply:
Yes, they are, in a sense that they are _typed_, and I think yes, they should.
Groups are not arbitrary strings of character, they are a predifined set of values.

At the moment I only have support for 1 group, and having it "hard-coded" allows for a more _sound_ error handling. Types are cool.

Reply:
So, if I want to add my own group I should add it to an enum? Sounds `suckless` to me :)

Reply:
The "group" is not a useful abstraction outside this project. There is no special processing or any other reason to reuse this struct outside of my binaries.

So yes, if one wishes to add support for a new group, they need to change the enum, and provide necessary functionality for all the code that uses this enum - most notably, update trait implementations and such.

I do not see a good reason or a practical solution to separate support for different groups, making them extensible.

Reply:
So then why do you need groups in the first place? If it is? technically, a constant. I feel dumb.

Reply:
It's a good question, actually. It allows me to brag about all of the cool things I learned during this bootcamp!

Group information is a part of database schema, which made it necessary for me to interact with it. Which brought the question of "_How should I interact with this information?_." And I decided on the answer "**Using types**."

In words of a popular article, _**parse don't validate**_. Not any string is a valid group. In multiple points in my application I need to know that the provided group is valid. To store that information, I use a type - aka "_Type Drive Development_". I think this sort of decision is idiomatic within Rust ecosystem.

Reply:
But what if you have a type, that parses the group when created, so then you always know that your group is valid when you have a value of this type. I think this is what that article is talking about. Your type can store a string value, but when you create a value you validate that this name.

Reply:
I see what you mean. Hmm.

I would need to derive a common format from the group names, and use it to parse groups generically. In this sense, a group would be not a member of a predefined set, but a string that has some properties. A quick glance at the group names from the original schedule suggests that they all follow `\w+(-\d+)?` regex structure.

This approach would provide support for all groups simultaneously. I think I like it.

Thanks for the suggestion, I'll keep it in mind for further refactoring.